CONTRIBUTING.md

```diff
@@ -1,76 +1,401 @@
+you can pick one to split into loose objects using `cat .git/objects/pack/[id].pack | git unpack-objects`;
+delete the `[id].pack` and `[id].idx` files after this. You can count the number of loose
+bjects using `ls .git/objects/??/* | wc -l`.
+
+**Deep history:** Usually large repositories also have deep histories, but you can use the
+[test-many-commits-1m repo](https://github.com/cirosantilli/test-many-commits-1m/) to
+target deep histories without the overhead of many objects. One issue with this repository:
+there are no merge commits, so you will need to use a different repository to test a "wide"
+commit history.
+
+**Large Index:** You can generate a large index and repo by using the scripts in
+`t/perf/repos`.  There are two scripts. `many-files.sh` which will generate a repo with
+same tree and blobs but different paths.  Using `many-files.sh -d 5 -w 10 -f 9` will create
+a repo with ~1 million entries in the index. `inflate-repo.sh` will use an existing repo
+and copy the current work tree until it is a specified size.
+
+Test Your Changes on Linux
+--------------------------
+
+It can be important to work directly on the [core Git codebase](https://github.com/git/git),
+such as a recent commit into the `master` or `next` branch that has not been incorporated
+into Git for Windows. Also, it can help to run functional and performance tests on your
+code in Linux before submitting patches to the Linux-focused mailing list.
+
+### Using the Windows Subsystem for Linux
+
+The [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
+allows you to [install Ubuntu Linux as an app](https://www.microsoft.com/en-us/store/p/ubuntu/9nblggh4msv6)
+that can run Linux executables on top of the Windows kernel. Internally,
+Linux syscalls are interpreted by the WSL, everything else is plain Ubuntu.
+
+First, open WSL (either type "Bash" in Cortana, or execute "bash.exe" in a CMD window).
+Then install the prerequisites, and `git` for the initial clone:
+
+```
+sudo apt-get update
+sudo apt-get install git gcc make libssl-dev libcurl4-openssl-dev libexpat-dev tcl tk gettext git-email
+```
+
+Then, clone and build:
+
+```
+git clone https://github.com/git-for-windows/git
+cd git
+git remote add -f upstream https://github.com/git/git
+make
+```
+
+Be sure to clone into `/home/[user]/` and not into any folder under `/mnt/?/` or your build
+will fail due to colons in file names.
+
+### Using a Linux Virtual Machine with Hyper-V
+
+If you prefer, you can use a virtual machine (VM) to run Linux and test your changes in the
+full environment. I like this setup, particularly because the test suite runs a lot faster
+on Linux than on Windows or with the WSL. I connect to the VM using an SSH terminal like
+[PuTTY](http://www.putty.org/). I've tried to make the following steps be as clear as
+possible, without missing any steps.
+
+* [Download an Ubuntu Server ISO](https://www.ubuntu.com/download/server).

```

Comment:
Personally, I just use the _Client_ version of Ubuntu (variant), because it's faster to install - a lot less menus in the installation process. Does not install `sshd` during setup. I chose Lubuntu to minimize the GUI overhead.

Reply:
Funnily enough, I personally like to `ssh` into my VM so that I can stay within the console windows on my desktop. So I guess your mileage may vary.

Reply:
I'm leaving this as-is, because the instructions I know are for the server variation, and the intent is to have no GUI interaction on the Linux VM.